platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonWriter.kt

```diff
@@ -29,6 +29,13 @@ class JsonWriter(private val writer: Writer) {
         endObject()
     }
 
+    private fun jsonListItem(body: () -> Unit) {
+        elementSeparator()
```

Comment:
❓ Is the `elementSeparator()` call in `jsonObject` still needed? It doesn't seem to make sense anymore.

Reply:
That's the actual fix in the PR. Without it, this call

    jsonList(listOf("first", "second").iterator()) { i -> property(i, i) }

yields this (invalid) json:

    ["first":"first""second":"second"]

Reply:
You are right. Good catch, thank you for taking a deeper look!

Reply:
I fixed that and did some code cleanup. See my commit messages for the details.

Reply:
In fact, removing `elementSeparator()` will result in invalid json generation. Without it, `JsonModelWriterTest` produces the following output:

```json
{
    "diagnostics": [
        {
            "trace": [
                {
                    "kind": "Unknown"
                }
            ],
            "input": [
                {
                    "name": ""
                }
            ]
        }{
            "trace": [
                {
                    "kind": "Unknown"
                }
            ],
            "input": [
                {
                    "name": ""
                }
            ]
        }
    ]
}

```

Reply:
That seems to be more of a problem with the test or with `DecoratedReportedProblemJsonSource` than with the implementation of `jsonObject`. The test is ends calling `writeToJson` which just calls `jsonObject` twice with no separators in between, is that the expected/legal usage? What if it's later changed to call `jsonObjectList` with no separators in between?

Reply:
You are absolutely right. In fact the issue I was seeing is indeed due to the illegal usage you described. 
Thanks for thinking about this deeply.